File: Assets/Scripts/Game/SaveSystem.cs
namespace HexaMerge.Game
{
    using HexaMerge.Core;
    using UnityEngine;
    using System.Collections.Generic;

    [System.Serializable]
    public class GameSaveData
    {
        public int score;
        public List<CellSaveData> cells = new List<CellSaveData>();
    }

    [System.Serializable]
    public class CellSaveData
    {
        public int q;
        public int r;
        public int value;
    }

    public static class SaveSystem
    {
        private const string SAVE_KEY = "GameSave";

        public static void Save(HexGrid grid, int score)
        {
            var data = new GameSaveData { score = score };

            foreach (var coord in grid.AllCoords)
            {
                var cell = grid.GetCell(coord);
                if (!cell.IsEmpty)
                {
                    data.cells.Add(new CellSaveData
                    {
                        q = coord.q,
                        r = coord.r,
                        value = cell.TileValue
                    });
                }
            }

            string json = JsonUtility.ToJson(data);
            PlayerPrefs.SetString(SAVE_KEY, json);
            PlayerPrefs.Save();
        }

        public static GameSaveData Load()
        {
            if (!PlayerPrefs.HasKey(SAVE_KEY))
                return null;

            string json = PlayerPrefs.GetString(SAVE_KEY);
            if (string.IsNullOrEmpty(json))
                return null;

            try
            {
                return JsonUtility.FromJson<GameSaveData>(json);
            }
            catch
            {
                return null;
            }
        }

        public static bool HasSave()
        {
            return PlayerPrefs.HasKey(SAVE_KEY);
        }

        public static void DeleteSave()
        {
            PlayerPrefs.DeleteKey(SAVE_KEY);
            PlayerPrefs.Save();
        }

        public static void ApplyToGrid(GameSaveData data, HexGrid grid)
        {
            foreach (var coord in grid.AllCoords)
            {
                grid.GetCell(coord).Clear();
            }

            foreach (var cellData in data.cells)
            {
                var coord = new HexCoord(cellData.q, cellData.r);
                var cell = grid.GetCell(coord);
                if (cell != null)
                {
                    cell.SetValue(cellData.value);
                }
            }
        }
    }
}

File: src/game/SaveSystem.js
/**
 * @fileoverview Game state persistence using localStorage and JSON.
 * Ported from Unity C# SaveSystem.cs to pure JavaScript ES Module.
 * Uses localStorage instead of PlayerPrefs, JSON.parse/stringify instead of JsonUtility.
 */

import { HexCoord } from '../core/HexCoord.js';

const SAVE_KEY = 'hexamerge_save';

/**
 * @typedef {Object} CellSaveData
 * @property {number} q
 * @property {number} r
 * @property {number} value
 */

/**
 * @typedef {Object} GameSaveData
 * @property {number} score
 * @property {CellSaveData[]} cells
 */

/**
 * Save the current grid state and score to localStorage.
 * Only non-empty cells are stored.
 * @param {import('../core/HexGrid.js').HexGrid} grid
 * @param {number} score
 */
export function save(grid, score) {
    /** @type {GameSaveData} */
    const data = {
        score,
        cells: []
    };

    const allCoords = grid.allCoords;
    for (let i = 0; i < allCoords.length; i++) {
        const coord = allCoords[i];
        const cell = grid.getCell(coord);
        if (cell && !cell.isEmpty) {
            data.cells.push({
                q: coord.q,
                r: coord.r,
                value: cell.value
            });
        }
    }

    try {
        const json = JSON.stringify(data);
        localStorage.setItem(SAVE_KEY, json);
    } catch (e) {
        console.warn('[SaveSystem] Failed to save game:', e);
    }
}

/**
 * Load saved game data from localStorage.
 * @returns {GameSaveData|null} Parsed save data, or null if no save exists or parsing fails.
 */
export function load() {
    try {
        const json = localStorage.getItem(SAVE_KEY);
        if (!json) return null;

        const data = JSON.parse(json);
        if (!data || typeof data.score !== 'number' || !Array.isArray(data.cells)) {
            return null;
        }

        return data;
    } catch (e) {
        console.warn('[SaveSystem] Failed to load game:', e);
        return null;
    }
}

/**
 * Check whether a save exists in localStorage.
 * @returns {boolean}
 */
export function hasSave() {
    try {
        return localStorage.getItem(SAVE_KEY) !== null;
    } catch (e) {
        return false;
    }
}

/**
 * Delete the save data from localStorage.
 */
export function deleteSave() {
    try {
        localStorage.removeItem(SAVE_KEY);
    } catch (e) {
        console.warn('[SaveSystem] Failed to delete save:', e);
    }
}

/**
 * Apply loaded save data to a grid.
 * Clears all cells first, then sets values from the save data.
 * @param {GameSaveData} data
 * @param {import('../core/HexGrid.js').HexGrid} grid
 */
export function applyToGrid(data, grid) {
    // Clear all cells first
    const allCoords = grid.allCoords;
    for (let i = 0; i < allCoords.length; i++) {
        const cell = grid.getCell(allCoords[i]);
        if (cell) cell.clear();
    }

    // Apply saved cell values
    for (let i = 0; i < data.cells.length; i++) {
        const cellData = data.cells[i];
        const coord = new HexCoord(cellData.q, cellData.r);
        const cell = grid.getCell(coord);
        if (cell) {
            cell.setValue(cellData.value);
        }
    }
}
